insights/parsers/samba_logs.py

```diff
@@ -0,0 +1,74 @@
+"""
+samba logs - files matching ``/var/log/samba/*.log``
+===================================================
+"""
+
+from .. import parser, LogFileOutput
+from datetime import datetime
+from insights.specs import Specs
+
+
+def strip_surrounds(s):
+    start = 0
+    end = len(s) - 1
+    while (s[start] == '[' and s[end] == ']') or \
+      (s[start] == '[' and s[end] == ',') or \
+      (s[start] == '(' and s[end] == ')'):
+        start += 1
+        end -= 1
+    if s[end] == ',':
+        end -= 1
+    return s[start:(end + 1)]
+
+
+@parser(Specs.samba_logs)
+class SAMBALog(LogFileOutput):
+    """
+    Parser class for reading samba log files.  The main work is done by the
+    LogFileOutput super-class.
+    .. note::
+        Please refer to its super-class :class:`insights.core.LogFileOutput`
+    Sample input::
+       [2018/12/07 07:09:44.812154,  5, pid=6434, effective(0, 0), real(0, 0)] ../source3/param/loadparm.c:1344(free_param_opts)
+        Freeing parametrics:
+       [2018/12/07 07:09:44.812281,  3, pid=6434, effective(0, 0), real(0, 0)] ../source3/param/loadparm.c:547(init_globals)
+        Initialising global parameters
+       [2018/12/07 07:09:44.812356,  2, pid=6434, effective(0, 0), real(0, 0)] ../source3/param/loadparm.c:319(max_open_files)
+        rlimit_max: increasing rlimit_max (1024) to minimum Windows limit (16384)
+       [2019/45/899 11:11:04.911891,  3, pid=15822, effective(0, 0), real(0, 0)] ../source3/printing/queue_process.c:236
+        (bq_sig_hup_handler) Reloading pcap cache after SIGHUP.
+    Each line is parsed into a dictionary with the following keys:
+        * **timestamp** - the date of the log line (as a string)
+        * **datetime** - the date as a datetime object (if conversion is possible)
+        * **pid** - process id of samba process being run
+        * **function** - the function within the module
+        * **message** - the body of the message
+        * **raw_message** - the raw message before being split.
+    Examples:
+        >>> logs = shared[SAMBALog]
```

Comment:
The `shared` is no longer used, please remove if from the `Examples`

Reply:
And you may want to test the lines in the `Examples` with the `doctest` of python, for example https://github.com/RedHatInsights/insights-core/blob/39ae8a265739170748fc9f81ad20d63751becae3/insights/parsers/tests/test_ksmstate.py#L43-L46

Reply:
@xiangce I wrote function to test docstring:
def test_samba_logs_doc():
    env = {'samba_logs': SAMBALog(context_wrap(STATE0))}
    failed, total = doctest.testmod(samba_logs, globs=env)
    assert failed == 0

Then I modified docstring:
"""
...
    Examples:

        >>> logs = shared[SAMBALog]

        >>> hups = logs.get("SIGHUP")

        >>> print('hups'+str(hups))
        >>> print(hups[0]['module'])
        'bq_sig_hup_handler'
...
"""

But pytest cannot find `shared`.
Failed example:
    logs = shared[SAMBALog]
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib64/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest insights.parsers.samba_logs.SAMBALog[0]>", line 1, in <module>
        logs = shared[SAMBALog]
    NameError: name 'shared' is not defined

Where is `shared` defined?

Reply:
`shared` is an old feature of insights-core to get the passed component, but it's deprecated already. You can remove it and test the parser in this way:

https://github.com/RedHatInsights/insights-core/blob/15c4fede4139c06ca2fb2aa090e763b08fc70d75/insights/parsers/ksmstate.py#L49-L52

Reply:
@xiangce You provided a example of doctest of a class which parses a file `/sys/kernel/mm/ksm/run`. Can you provide example for doctest of class parsing a log file?
Also, if you can suggest what would be suitable doctest string for current parser?

Reply:
any updates?

Reply:
@xiangce Any updated?

Reply:
The doctests are basically the same for either log file or non-logfile, but here is an example for you
https://github.com/RedHatInsights/insights-core/blob/045899f992325753530ec0ac299e133bf50f5516/insights/parsers/tests/test_ceph_log.py#L30-L33

Sorry for the late.

Reply:
@xiangce Thanks will look into

Reply:
@xiangce This is doctest function:
`
def test_samba_logs_doc():
    env = {"samba_logs": SAMBALog(context_wrap(SAMBA_LOG_NORMAL)),}
    failed, total = doctest.testmod(samba_logs, globs=env)
    assert failed == 0
`

And these tests in docstring it passes:
`
    Examples:
        >>> 'Fake' in samba_logs
        True
        >>> 'pid=15822, effective(0, 0), real(0, 0)]' in samba_logs 
        True
        >>> len(samba_logs.get('Fake line')) == 1
        True
`

Reply:
@amitkumar50 - the code and test result you attached look good to me. Would you please commit it to this PR?

Reply:
please fix the following error in test  by importing the doctest at the top
```
>       failed, total = doctest.testmod(samba_logs, globs=env)
E       NameError: global name 'doctest' is not defined
```

BTW, please fix the DCO check error as well.
- https://github.com/RedHatInsights/insights-core/pull/2362/checks?check_run_id=372068611

Reply:
||commit it to this PR
do you mean rebase over master and again commit it?

Reply:
@xiangce

Reply:
@xiangce fixed DCO check error